samples/template-operator/operator/controllers/sample_controller.go

```diff
@@ -61,13 +72,26 @@ func (r *SampleReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctr
 		return ctrl.Result{}, client.IgnoreNotFound(err)
 	}
 
+	// check if deletionTimestamp is set, retry until it gets fully deleted
+	if !sampleResource.DeletionTimestamp.IsZero() &&
+		sampleResource.Status.State != v1alpha1.SampleStateDeleting {
+		// if the status is not yet set to deleting, also update the status
+		sampleResource.Status.State = v1alpha1.SampleStateDeleting
+		return ctrl.Result{}, r.Status().Update(ctx, sampleResource)
+	}
+
+	// add deletion finalizer
+	if controllerutil.AddFinalizer(sampleResource, deletionFinalizer) {
+		return ctrl.Result{}, r.Client.Update(ctx, sampleResource)
+	}
+
 	switch sampleResource.Status.State {
 	case "":
 		return ctrl.Result{}, r.HandleInitialState(ctx, sampleResource)
 	case v1alpha1.SampleStateProcessing:
-		return ctrl.Result{}, r.HandleProcessingState(ctx, sampleResource)
+		return ctrl.Result{}, r.HandleProcessingState(ctx, sampleResource, &logger)
```

Comment:
the logger you do not need to pass, you can just call `logger := log.FromContext(ctx)`